src/messaging/ExchangeContext.cpp

```diff
@@ -77,14 +78,55 @@ void ExchangeContext::SetResponseTimeout(Timeout timeout)
 
 CHIP_ERROR ExchangeContext::SendMessage(uint16_t protocolId, uint8_t msgType, PacketBufferHandle msgBuf,
                                         const SendFlags & sendFlags)
+{
+    CHIP_ERROR err                         = CHIP_NO_ERROR;
+    Transport::PeerConnectionState * state = nullptr;
+
+    VerifyOrReturnError(mExchangeMgr != nullptr, CHIP_ERROR_INTERNAL);
+
+    state = mExchangeMgr->GetSessionMgr()->GetPeerConnectionState(mSecureSession);
+    VerifyOrReturnError(state != nullptr, CHIP_ERROR_NOT_CONNECTED);
+
+    // If a group message is to be transmitted to a destination node whose message counter is unknown.
+    if (ChipKeyId::IsAppGroupKey(state->GetLocalKeyID()) && !state->isPeerMsgCounterSynced())
+    {
+        MessageCounterSyncMgr * messageCounterSyncMgr = mExchangeMgr->GetMessageCounterSyncMgr();
+        VerifyOrReturnError(messageCounterSyncMgr != nullptr, CHIP_ERROR_INTERNAL);
+
+        // Queue the message as need for sync with destination node.
+        err = messageCounterSyncMgr->AddToRetransmissionTable(protocolId, msgType, sendFlags, std::move(msgBuf), this);
+        ReturnErrorOnFailure(err);
+
+        // Initiate message counter synchronization if no message counter synchronization is in progress.
+        if (!messageCounterSyncMgr->IsMsgCounterSyncReqInProgress())
```

Comment:
AFAIK, `IsMsgCounterSyncReqInProgress` is a per-session state, not a global singleton state, I'm not sure how do you get it from a singleton manager.

Reply:
Currently, any new group messages need message counter sync will be queued for later processing if Msg Counter Sync is in progress. The delay introduced by the back logging might be ok considering MCSP is rarely triggered. 

I modified the logic to track this status per-connection. Assume we have multiple group messages targeted to same Peer come at the same time, and all of them need to trigger message counter sync process, we don't need to start three MCSP sessions in this case, since the result could be shared. 

If we have multiple group messages targeted to different Peer, current design allow them to run MCSP simultaneously.